clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +69,729 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+        for info in infos {
+            let node = body.source_scopes[info.scope]
+                .local_data
+                .as_ref()
+                .assert_crate_local()
+                .lint_root;
+            span_lint_hir(cx, REDUNDANT_CLONE, node, info.span, "redundant clone");
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct Value {
+    // Two values which were last modified at different locations may contain different values.
+    // Joining two such values needs to be treated as a modification.
+    last_modified: Option<Location>,
+}
+
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+struct CloneDomain {
+    // Values which are clones of each other and may still contain the same value.
+    links: FxHashSet<(Local, Local, BasicBlock)>,
+    // Values which had their linked value modified.
+    // Reading one of these values means the clone call is required.
+    diverged_clones: FxHashSet<(Local, BasicBlock)>,
+    values: FxHashMap<Local, Value>,
+}
+impl CloneDomain {
+    fn mark_modified(&mut self, local: Local) -> bool {
+        let mut changed = false;
+        for &(l1, l2, block) in &self.links {
+            let local = if l1 == local {
+                l2
+            } else if l2 == local {
+                l1
+            } else {
                 continue;
+            };
+            changed |= self.diverged_clones.insert((local, block));
+        }
+        changed
+    }
+}
+impl JoinSemiLattice for CloneDomain {
+    fn join(&mut self, other: &Self) -> bool {
+        let prev_len = self.links.len();
+        self.links.extend(&other.links);
+        let mut changed = prev_len != self.links.len();
+
+        let prev_len = self.diverged_clones.len();
+        self.diverged_clones.extend(&other.diverged_clones);
+        changed |= prev_len != self.diverged_clones.len();
+
+        for (&local, value) in &other.values {
+            match self.values.entry(local) {
+                Entry::Vacant(e) => {
+                    changed = true;
+                    e.insert(value.clone());
+                },
+                Entry::Occupied(mut e) => {
+                    if e.get().last_modified.is_some() && e.get().last_modified != value.last_modified {
+                        e.get_mut().last_modified = None;
+                        changed = true;
+                    }
+                    if e.get().last_modified.is_none() {
+                        changed |= self.mark_modified(local);
+                    }
+                },
             }
+        }
+        changed
+    }
+}
+impl DebugWithContext<CloneAnalysis<'_, '_, '_>> for CloneDomain {}
+
+struct CloneAnalysis<'a, 'mir, 'tcx> {
+    body: &'mir Body<'tcx>,
+    tracked: BitSet<Local>,
+    clone_sources: &'a FxHashMap<BasicBlock, (Local, HybridBitSet<Local>)>,
+    link_buf: Vec<(Local, Local, BasicBlock)>,
+    borrowers: PossibleBorrowerMap<'mir, 'tcx>,
+    required_clones: FxHashSet<BasicBlock>,
```

Comment:
This could be calculated in a `ResultsVisitor` thereby resolving the earlier issue with `PossibleBorrowerMap`, but that requires the visitor to detect reads and moves the same way the analysis does.

Given that handling each read is a small amount of work, and using a visitor requires a whole extra pass over every block it's probably better to do it this way. It's also just less work to implement/maintain.